Perf guidelines (#303)
MicrosoftDocs/OData-docs: Odata-docs/webapi-8/tutorials/using-odatautf8jsonwriter-to-improve-serialization-performance.md

---
title:  "Using ODataUtf8JsonWriter to improve serialization performance"
description: Learn how to configure ASP.NET Core OData 8 to use Utf8JsonWriter to improve serialization performance. 
date:   2023-08-11
ms.date: 08/11/2023
author: habbes
ms.author: clhabins
---

# Using ODataUtf8JsonWriter to improve serialization performance

**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> `Microsoft.OData.Core` version 7.12.2 introduced a new JSON writer that writes directly to a `Stream` rather than using a `TextWriter`. It also introduced a default implementation of this "stream-based" JSON writer that’s based on .NET’s [`Utf8JsonWriter`](/dotnet/api/system.text.json.utf8jsonwriter). Internally, it's referred to as `ODataUtf8JsonWriter`. This writer offers better performance than the default JSON writer used in `Microsoft.OData.Core`. To learn more about the `ODataUtf8JsonWriter`, [visit this page](../../odatalib/v7/using-utf8jsonwriter-for-better-performance.md). In this article, we are going to show you how to configure your ASP.NET Core OData application to use the `ODataUtf8JsonWriter`.

Reviewer: > Internally, it's referred to as `ODataUtf8JsonWriter`

 I don't think the above statement is factual or clear. It sounds more like you're saying `ODataUtf8JsonWriter` is a nickname for `Utf8JsonWriter`. Maybe you could say something like:

> The OData Core library implements a wrapper on top of the `Utf8JsonWriter` named `ODataUtf8JsonWriter`

> This feature is available in ASP.NET Core OData 8 starting from v8.0.11.

Reviewer: ```suggestion
This feature is available in ASP.NET Core OData version 8.0.11 or later.
```

> You can configure custom services when registering new OData route components using the `AddRouteComponents` method. In this case, you need to
> inject a service that implements `IStreamBasedJsonWriterFactory`. The `Microsoft.OData.Core` library provides a default implementation of this interface: `DefaultStreamBasedJsonWriterFactory.Default`.

Reviewer: ```suggestion
You can register OData services through use of `AddRouteComponents` method. To use the `ODataUtfJsonWriter`, register the `IStreamBasedJsonWriterFactory` service. The `Microsoft.OData.Core` library provides a default implementation of this interface - `DefaultStreamBasedJsonWriterFactory.Default`.
```

> Here's an example of how to add this to your application:

Reviewer: ```suggestion
Here's how you can do this in your application:
```

> In the file containing your application start up code, import the `Microsoft.OData.Json` namespace and register the `IStreamBasedJsonWriterFactory` service as follows:

Reviewer: ```suggestion
In your application startup code, import the `Microsoft.OData.Json` namespace and register the `IStreamBasedJsonWriterFactory` service as follows:
```

```c#
services.AddControllers()
    .AddOData(options =>
    options.AddRouteCompontents("odata", model, services =>
    {
        services.AddSingleton<IStreamBasedJsonWriterFactory>(_ => DefaultStreamBasedJsonWriterFactory.Default);
    });
```

> That's all that's required to substitute the default JSON writer with `ODataUtf8JsonWriter`.

Reviewer: ```suggestion
That is all that's required to substitute the default JSON writer with `ODataUtf8JsonWriter`.
```
☝️  Microsoft Word's suggestion

---
title:  "Using ODataUtf8JsonWriter to improve serialization performance"

> description: Learn how to configure ASP.NET Core OData 8 to use Utf8JsonWriter to improve serialization performance. 

Reviewer: ```suggestion
description: Learn how to configure ASP.NET Core OData 8 to use ODataUtf8JsonWriter to improve serialization performance. 
```

date:   2023-08-11
ms.date: 08/11/2023
author: habbes
ms.author: clhabins
---

# Using ODataUtf8JsonWriter to improve serialization performance

**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> `Microsoft.OData.Core` version 7.12.2 introduced a new JSON writer that writes to a `Stream` instead of a `TextWriter`. It also introduced a default implementation of this "stream-based" JSON writer that’s based on .NET’s [`Utf8JsonWriter`](/dotnet/api/system.text.json.utf8jsonwriter). Internally, it's referred to as `ODataUtf8JsonWriter`. This writer is more performant than the default JSON writer used in `Microsoft.OData.Core`. To learn more about the stream-based writer, [visit this page](../../odatalib/v7/using-utf8jsonwriter-for-better-performance.md). In this article, we are going to show you how to configure your ASP.NET Core OData application to use the `ODataUtf8JsonWriter`.

Reviewer: ```suggestion
`Microsoft.OData.Core` version 7.12.2 introduced a new JSON writer that writes directly to a `Stream` rather than through a `TextWriter`. It also introduced a default implementation of this "stream-based" JSON writer that’s based on .NET’s [`Utf8JsonWriter`](/dotnet/api/system.text.json.utf8jsonwriter). Internally, it's referred to as `ODataUtf8JsonWriter`. This writer offers better performance than the default JSON writer used in `Microsoft.OData.Core`. To learn more about the stream-based writer, [visit this page](../../odatalib/v7/using-utf8jsonwriter-for-better-performance.md). In this article, we are going to show you how to configure your ASP.NET Core OData application to use the `ODataUtf8JsonWriter`.
```
Be careful in your use of "stream-based". All writers ultimately write to a stream. The main differentiation is that the `ODataUtf8JsonWriter` writes directly to the stream and rides on the .NET `Utf8JsonWriter`

> `Microsoft.OData.Core` version 7.12.2 introduced a new JSON writer that writes to a `Stream` instead of a `TextWriter`. It also introduced a default implementation of this "stream-based" JSON writer that’s based on .NET’s [`Utf8JsonWriter`](/dotnet/api/system.text.json.utf8jsonwriter). Internally, it's referred to as `ODataUtf8JsonWriter`. This writer is more performant than the default JSON writer used in `Microsoft.OData.Core`. To learn more about the stream-based writer, [visit this page](../../odatalib/v7/using-utf8jsonwriter-for-better-performance.md). In this article, we are going to show you how to configure your ASP.NET Core OData application to use the `ODataUtf8JsonWriter`.

Reviewer: In the redirection file, you the path with `v7` is routed to the path without. I think you should use the path without `v7` here.

Author: I think the redirection file was about the article in ODL, not Web API.

> This feature is available in ASP.NET Core OData 8 starting from v8.0.11, and only available in .NET Core 3.1 and later (i.e. .NET 6+). It is not available on .NET Framework.

Reviewer: ```suggestion
This feature is available in ASP.NET Core OData 8 starting from version 8.0.11. It is not available on .NET Framework.
```
Since ASP.NET Core OData 8 cannot be used with versions of .NET Core lower than .NET Core 3.1, I think we can drop that part of the statement altogether

You can configure custom services when registering new OData route components using the `AddRouteComponents` method. In this case, you need to
inject a service that implements `IStreamBasedJsonWriterFactory`. The `Microsoft.OData.Core` library provides a default implementation of this interface: `DefaultStreamBasedJsonWriterFactory.Default`.

Here's an example of how to add this to your application:

> Add the following using statement to the `Startup.cs` file, or the file where you configure application services:
> 
> ```c#
> using Microsoft.OData.Json;
> ```
> 
> Then register the new factory inside the Startup.ConfigureServices method or the method where you’re configuring application services if not Startup.ConfigureServices:

Reviewer: ```suggestion
In the file containing the application start up code, import the `Microsoft.OData.Json` namespace and register the `IStreamBasedJsonWriterFactory` service as follows:
```

```c#
services.AddControllers()
    .AddOData(options =>
    options.AddRouteCompontents("odata", model, services =>
    {
        services.AddSingleton<IStreamBasedJsonWriterFactory>(_ => DefaultStreamBasedJsonWriterFactory.Default);
    });
```

> That's it. This is all the configuration that's needed to start using the new writer.

Reviewer: ```suggestion
That's the only configuration needed to start using the `ODataUtf8JsonWriter`.
```
